scanner/enricher/nvd/nvd.go

```diff
@@ -245,11 +245,13 @@ func (e *Enricher) FetchEnrichment(ctx context.Context, _ driver.Fingerprint) (i
 					_ = f.Close()
 				}()
 				dec := json.NewDecoder(f)
-				var vuln *schema.CVEAPIJSON20DefCVEItem
-				for err := dec.Decode(&vuln); err == nil; err = dec.Decode(vuln) {
-					if !yield(vuln) {
+				var vuln schema.CVEAPIJSON20DefCVEItem
+				for err = dec.Decode(&vuln); err == nil; err = dec.Decode(&vuln) {
```

Comment:
I wonder why it never failed with any type error or something before. Anyways, why do we call decode twice?

Reply:
Huh, nice catch.

Maybe just moving to `for {`; then you can declare a new empty `vuln` at every iteration.

```
for {
	var vuln schema.CVEAPIJSON20DefCVEItem
	err = dec.Decode(&vuln)
	if err != nil {
		break
	}
	if !yield(&vuln) {
		break
	}
}

Reply:
> I wonder why it never failed with any type error or something before. Anyways, why do we call decode twice?

When calling `Decode` the passed in argument does not have to be empty, as long as the type is correct. When the JSON is decoded into the type, it just fills out the fields which were set in the JSON and leaves any fields already set in the passed in argument as-is. `Decode` is not quite called twice. It's more like we initialize the loop with a call to `Decode` and upon next iteration call it again. Same idea as `for i := 0; i < 10; i++`